project_report/project_report_latex.py

```diff
@@ -0,0 +1,502 @@
+        return new_rows
+
+    @staticmethod
+    def create_vertical_table(container, header, rows, column_def=None, footer=None):
+        def add_footer_rows(foot):
+            if not isinstance(foot, list):
+                foot = [foot]
+            for f in foot:
+                data_table.add_row((MultiColumn(ncol - 1, align='l', data=f), ''))
+
+        ncol = len(header)
+        if not column_def:
+            column_def = ' '.join(['X[r]'] * ncol)
+        # Using the tabu package -> http://mirrors.ibiblio.org/CTAN/macros/latex/contrib/tabu/tabu.pdf
+        with container.create(LongTabu(column_def, width=ncol)) as data_table:
+            data_table.add_hline()
+            data_table.add_row(header, mapper=bold)
+            data_table.add_hline()
+            data_table.end_table_header()
+            # Footer contains the next page notice
+            data_table.add_hline()
+            if footer:
+                add_footer_rows(footer)
+            data_table.add_row((MultiColumn(ncol, align='r', data='Continued on Next Page'),))
+            data_table.add_hline()
+            data_table.end_table_footer()
+            # Last footer does not have the next page row
+            if footer:
+                add_footer_rows(footer)
+            data_table.end_table_last_footer()
+            for r in rows:
+                data_table.add_row(r)
+            data_table.add_hline()
+
+    @staticmethod
+    def create_horizontal_table(container, rows):
+        """ Meant to be used for only two columns where the header is the first column"""
+        # Convert cell containing lists into multilines cells
+        converted_rows = []
+        for row in rows:
+            converted_row = []
+            for cell in row:
+                if isinstance(cell, list):
+                    converted_row.append('\n'.join(cell))
+                else:
+                    converted_row.append(cell)
+            converted_rows.append(converted_row)
+
+        column_def = r'>{\bfseries}lX[l]'
+        with container.create(Tabu(column_def, row_height=1.6)) as data_table:
+            for r in converted_rows:
+                data_table.add_row(r)
+
+    @staticmethod
+    def first_pages_style():
+        # Generating first page style
+        page = PageStyle("firstpage")
+
+        # Address in small print in footer
+        address = 'Edinburgh Genomics Clinical, The Roslin Institute Easter Bush, Midlothian EH25 9RG Scotland, UK'
+        with page.create(Foot("C")) as footer:
+            footer.append(FootnoteText(address))
+
+        return page
+
+    @staticmethod
+    def all_pages_style(title):
+        # Generating report page style
+        page = PageStyle("allpages")
+
+        # EG logo in header
+        with page.create(Head("L")) as header_left:
+            with header_left.create(MiniPage(pos='c', align='l')) as logo_wrapper:
+                logo_wrapper.append(HRef(
+                    url='https://genomics.ed.ac.uk',
```

Comment:
Same with this URL, I would prefer if anything which could ever change is stored in a separate location and simply referenced in the source code.